File: C/COUNT_INVERSIONS.c
// COUNT INVERSIONS:
/*
Given an array of integers. Find the Inversion Count in the array.  Two elements arr[i] and arr[j] form an inversion if arr[i] > arr[j] and i < j.

Inversion Count: For an array, inversion count indicates how far (or close) the array is from being sorted. If the array is already sorted then the inversion count is 0.
If an array is sorted in the reverse order then the inversion count is the maximum. */




#include <stdio.h>
#include <stdlib.h>

long long count = 0;


void merge(long long arr[], long long left, long long middle, long long right)
{
    long long a1 = middle - left + 1;
    long long a2 = right - middle;
    long long i, j, k;

    
    long long *arr1 = (long long *)malloc(a1 * sizeof(long long));
    long long *arr2 = (long long *)malloc(a2 * sizeof(long long));

    
    for (i = 0; i < a1; i++)
        arr1[i] = arr[left + i];
    for (j = 0; j < a2; j++)
        arr2[j] = arr[middle + 1 + j];

    i = 0;
    j = 0;
    k = left;

    
    while (i < a1 && j < a2)
    {
        if (arr1[i] <= arr2[j])
        {
            arr[k] = arr1[i];
            i++;
        }
        else
        {
            arr[k] = arr2[j];
            j++;
            count += (a1 - i); 
        }
        k++;
    }

    
    while (i < a1)
    {
        arr[k] = arr1[i];
        i++;
        k++;
    }

    
    while (j < a2)
    {
        arr[k] = arr2[j];
        j++;
        k++;
    }

    free(arr1);
    free(arr2);
}


void mergeSort(long long arr[], long long left, long long right)
{
    if (left < right)
    {
        long long middle = left + (right - left) / 2;
        mergeSort(arr, left, middle);
        mergeSort(arr, middle + 1, right);
        merge(arr, left, middle, right);
    }
}


long long inversionCount(long long arr[], long long n)
{
    count = 0; 
    mergeSort(arr, 0, n - 1);
    return count;
}


int main()
{
    long long arr[] = {5, 3, 2, 4, 1};
    long long n = sizeof(arr) / sizeof(arr[0]);

    printf("Number of inversions: %lld\n", inversionCount(arr, n));
    return 0;
}

print(f "Number of inversions: {ms.inversion_count(arr)}")

File: python/COUNT_INVERSIONS.py
class MergeSort : 
    def  __init__(self) : 
        self.count = 0

    def merge(self, arr, left, middle, right) :
        arr1 = arr [left:middle + 1] 
        arr2 = arr [middle + 1:right + 1]

        i = j = 0 
        k = left


        while i < len(arr1) and j < len(arr2) : 
            if arr1[i] <= arr2[j] : 
                arr[k] = arr1[i] 
                i+= 1 
                
            else : 
                arr[k] = arr2[j] 
                j += 1 
                self.count += (len(arr1) - i) 
                k+= 1


        while i < len(arr1) : 
            arr[k] = arr1[i] 
            i+= 1 
            k+= 1


        while j < len(arr2) :
            arr[k] = arr2[j] 
            j+= 1 
            k+= 1

                                                               
    def merge_sort(self, arr, left, right) : 
        if left < right : 
            middle = left + (right - left) // 2
            self.merge_sort(arr, left, middle)
            self.merge_sort(arr, middle + 1, right)
            self.merge(arr, left, middle, right)

    def inversion_count(self, arr) : 
        self.count = 0 
        self.merge_sort(arr, 0, len(arr) - 1) 
        return self.count

